config/Makefile

```diff
@@ -0,0 +1,376 @@
+#
+define set-kustomize-image
+	$(if $(filter $(DEFAULT_IMAGE),$(IMAGE_NAME):$(TAG)),,\
+		@cd $(1) && $(KUSTOMIZE) edit set image $(DEFAULT_IMAGE)=$(IMAGE_NAME):$(TAG))
+endef
+
+#
+# Macro for editing kustomization to define
+# the namespace
+#
+# Parameter: directory of the kustomization.yaml
+#
+define set-kustomize-namespace
+	@cd $(1) && $(KUSTOMIZE) edit set namespace $(NAMESPACE)
+endef
+
+#
+# Macro for swapping a resource directory to an alternative directory
+#
+# Parameters:
+# * directory of the kustomization.yaml
+# * source directory
+# * target directory
+#
+define swap-rbac
+	@cd $(1) && \
+		$(KUSTOMIZE) edit remove resource ../$(2) && \
+		$(KUSTOMIZE) edit add resource ../$(3)
+endef
+
+#
+# Add or remove a patch on a kustomization.yaml
+# targetting the deployment resource
+#
+# Parameters:
+# * directory of the kustomization.yaml
+# * [add, remove]
+# * path of patch
+#
+define add-remove-deployment-patch
+	@cd $(1) && \
+		$(KUSTOMIZE) edit $(2) patch --path patches/$(3) --kind Deployment --name camel-k-operator
+endef
+
+#
+# Macro for adding / removing the prometheus resources for monitoring
+#
+define add-remove-operator-monitoring
+	cd $(1) && \
+		$(KUSTOMIZE) edit $(2) resource ../prometheus
+endef
+
+.PHONY: have-platform check_admin setup-cluster .setup-kubernetes .setup-openshift setup
+
+#
+# Determine the platform of the cluster based on
+# either the use of querying through a go-client
+# or using an installed client, ie. oc or kubectl
+#
+find-platform:
+ifndef PLATFORM
+PLATFORM=$(shell script/check_platform.sh)
+endif
+
+#
+# Checks if the cluster platform has been defined correctly either by the user
+# or by the platform_check script.
+#
+have-platform: find-platform
+ifeq ($(PLATFORM),openshift)
+	@echo Platform identified as 'openshift'
+else ifeq ($(PLATFORM),kubernetes)
+	@echo Platform identified as 'kubernetes'
+else
+	@echo "****"
+	@echo "**** ERROR: Cannot continue as cluster platform cannot be identified ****"
+	@echo "****"
+	@exit 1
+endif
+
+#
+# Checks if the cluster user has the necessary privileges to be a cluster-admin
+# In this case if the user can list the CRDs then probably a cluster-admin
+#
+check-admin: kubectl
+	@output=$$(kubectl get crd 2>&1) || (echo "****" && echo "**** ERROR: Cannot continue as user is not a Cluster-Admin ****" && echo "****"; exit 1)
+
+#
+# Setup the cluster installation by installing crds and cluster roles.
+#
+# Will either call setup-cluster-openshift (then setup-cluster-kubernetes) or
+# setup-cluster-kubernetes depending on the identity of the cluster
+#
+# PARAMETERS:
+#   NAMESPACE: Sets the namespace for the resources
+#   PLATFORM:  Override the discovered platform, if required
+#   DEBUG:     true - Prints the resources to be applied instead of applying them
+#
+setup-cluster: have-platform kustomize kubectl check-admin
+# Set the namespace in the setup-cluster kustomization yaml
+	@$(call set-kustomize-namespace,$@)
```

Comment:
Is this really needed?

Reply:
I think the culprit comes from the ClusterRoleBinding, that requires knowledge about the namespace where the operator is to be installed. I'd be inclined to think ClusterRoleBinding resources should be created during the setup, not the cluster setup, which is what the `kamel install` command does. I may see it's for the ability to install the operator as standard user, but in practice, installing the operator requires cluster admin. Plus, that requires to run the cluster setup each time you install the operator in different namespaces.

Reply:
I separated out `setup` and `cluster-setup` late in the day when implementing the e2e tests. I noted that `kamel install` has a `--cluster-setup` switch which only executes the cluster-level resource applications. Having them separate also mirrors the OLM installation procedure where cluster-level resources are applied by the OLM admin user.

`make setup` has the additional role of installing the contents of the `infrastructure` directory. This was something I wrestled with and current working practice is still not well laid out (see [here](https://github.com/operator-framework/operator-lifecycle-manager/issues/2063#issuecomment-811962533). The operator should take care of applying all resources required for its operands to function. Therefore, that implies any service accounts are the responsibility of the operator (since they have to be added to the namespace where the operands are being installed and not the operator ns). Therefore, the `infrastucture` directory, IMHO, represents those items that the operator should be installing rather than `kamel install`. 

So this argues for keeping them separate. However, your point about the ClusterRoleBindings is spot on. Fuse-Online escaped this by renaming the bindings with a namespace suffix but that is at best a 'sticking plaster'. I can refactor the ClusterRoleBindings into `make setup` but that will mandate that the user executing it has to be a cluster-admin. Are you good with that?

This will mean that `make cluster-setup` only does then have to be executed once. My only hesitation is that it will require a separate directory for the ClusterRoleBindings since the `make bundle` [kustomization.yaml](https://github.com/apache/camel-k/pull/2284/files#diff-fe545782dbfdb2c00c396b25f7a8990e1d4af2d68d02e25363dc54c837f8aac6R30) adds both ClusterRoles and ClusterRoleBindings to the CSV.

Reply:
I've quickly checked and it seems the operator are already granted all the roles that it would be bounding to the builder service account. So I think we could go for the option that the operator creates all the operand roles and bindings, which would make the `infrastructure` directory "embed resources" rather than "installation configuration". So we could move its content in the `deploy` directory (to be renamed `resources` as I've proposed in another comment above).

Reply:
Done